Fix add to dashboard after saving

neodescis opened the pull request:
### Description

When creating a new visualization from a dashboard, and saving it with the "Add to Dashboards after saving" toggle checked, adding the newly saved visualization to the dashboard can fail. This is easily reproducible when `storeInSessionStorage` is enabled.

The way it is supposed to work is that the scopedHistory receives the ID of the newly create visualization in the current location state, and the dashboard container uses that to add the visualization. However, when the dashboard application state is created, it can update the URL at startup, which causes the location state to be lost. This fix preserves the current scopedHistory location state across that URL update.

### Issues Resolved

- fix: Fix add to dashboard after saving 

## Screenshot

N/A

## Testing the changes

1. Enable `storeInSessionStorage`
2. Go to an existing dashboard and click Edit
3. Click Create New
4. Pick a visualization, make whatever changes are desired, then click Save
5. Ensure "Add to Dashboards after saving" is enabled in the modal, then click Save and Return
6. The new visualization should be added to the dashboard. This would fail before this change.

## Changelog
- fix: Preserve location state at dashboard app startup to fix adding a new visualization

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

abbyhu2000 commented:
Thanks for the fix! i have verified the fix is working properly.

@neodescis Do you think u could fix the changelog failure? It also seems like src/plugins/dashboard/public/application/utils/create_dashboard_app_state.test.tsx line 146 `describe('updateStateUrl', () => {...}` needs some updates. Can u resolve the error and also possibly add a unit test to test the new logics?

tungsheng commented:
@abbyhu2000 I can help resolve the test failure and add new tests! Let me know if that works. Thank you!! cc @sulabhjain-dev

ananzh commented:
@neodescis could you install https://github.com/apps/opensearch-changeset-bot which should help to fix the changelog error

ananzh commented:
@neodescis could you rebase based on the latest main and then push? I am not seeing `update branch` button or `re-run test` button which is wired.

neodescis commented:
> @neodescis could you rebase based on the latest main and then push? I am not seeing `update branch` button or `re-run test` button which is wired.

Done. I believe if the pull request comes from an organization-owned repo, those buttons go away for you here for some reason. It is weird indeed.

neodescis commented:
> @neodescis could you install https://github.com/apps/opensearch-changeset-bot which should help to fix the changelog error

Installed... let me know if there's something else.

neodescis commented:
> @abbyhu2000 @neodescis I can help resolve the test failure and add new tests! Let me know if that works. Thank you!! cc @sulabhjain-dev

From my perspective that would be great! I won't have time to get to it for a few days.

ananzh commented:
@neodescis seems for now only one unit test suite fails 
```
Summary of all failing tests
 FAIL  src/plugins/dashboard/public/application/utils/create_dashboard_app_state.test.tsx
  ● createDashboardGlobalAndAppState › encountered a declaration exception

    TypeError: Cannot read properties of undefined (reading 'location')

      161 |   // immediately forces scheduled updates and changes location
      162 |   // scoped history state is preserved to allow embeddable state transfer
    > 163 |   const previousState = scopedHistory.location.state;
          |                                       ^
      164 |   const changed = osdUrlStateStorage.flush({ replace });
      165 |   if (changed) {
      166 |     scopedHistory.replace({

      at updateStateUrl (src/plugins/dashboard/public/application/utils/create_dashboard_app_state.tsx:163:39)
      at createDashboardGlobalAndAppState (src/plugins/dashboard/public/application/utils/create_dashboard_app_state.tsx:136:3)
      at Suite.<anonymous> (src/plugins/dashboard/public/application/utils/create_dashboard_app_state.test.tsx:70:39)
      at Object.<anonymous> (src/plugins/dashboard/public/application/utils/create_dashboard_app_state.test.tsx:54:1)

  ● updateStateUrl › encountered a declaration exception

    TypeError: Cannot read properties of undefined (reading 'location')

      161 |   // immediately forces scheduled updates and changes location
      162 |   // scoped history state is preserved to allow embeddable state transfer
    > 163 |   const previousState = scopedHistory.location.state;
          |                                       ^
      164 |   const changed = osdUrlStateStorage.flush({ replace });
      165 |   if (changed) {
      166 |     scopedHistory.replace({

      at updateStateUrl (src/plugins/dashboard/public/application/utils/create_dashboard_app_state.tsx:163:39)
      at Suite.<anonymous> (src/plugins/dashboard/public/application/utils/create_dashboard_app_state.test.tsx:151:17)
      at Object.<anonymous> (src/plugins/dashboard/public/application/utils/create_dashboard_app_state.test.tsx:146:1)


Test Suites: 1 failed, 1 skipped, 583 passed, 584 of 585 total
Tests:       2 failed, 15 skipped, 3880 passed, 3897 total
Snapshots:   492 passed, 492 total
```

tungsheng commented:
> > @abbyhu2000 @neodescis I can help resolve the test failure and add new tests! Let me know if that works. Thank you!! cc @sulabhjain-dev
> 
> From my perspective that would be great! I won't have time to get to it for a few days.

@ananzh @abbyhu2000 @neodescis
I have created a new PR #9072 that cherry-picked the original commit and added unit tests. Also added original PR link in the description. Feel free to share any comments/suggestions, and I will keep improving accordingly. Thank you for your help!! 🙏 
cc @sulabhjain-dev

neodescis commented:
> > > @abbyhu2000 @neodescis I can help resolve the test failure and add new tests! Let me know if that works. Thank you!! cc @sulabhjain-dev
> > 
> > 
> > From my perspective that would be great! I won't have time to get to it for a few days.
> 
> @ananzh @abbyhu2000 @neodescis I have created a new PR #9072 that cherry-picked the original commit and added unit tests. Also added original PR link in the description. Feel free to share any comments/suggestions, and I will keep improving accordingly. Thank you for your help!! 🙏 cc @sulabhjain-dev

Thank you for pushing this forward for me!
